arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp

```diff
@@ -100,82 +98,39 @@ auto DocumentFollowerState::applyEntries(
-              auto res =
-                  data.core->ensureShard(doc.shardId, collectionId, doc.data);
-              if (res.fail()) {
-                LOG_CTX("75cda", FATAL, self->loggerContext)
-                    << "Failed to create shard " << doc.shardId
-                    << " of collection " << collectionId
-                    << " with error: " << res;
-                FATAL_ERROR_EXIT();
-              }
-              LOG_CTX("d82d5", TRACE, self->loggerContext)
-                  << "Created local shard " << doc.shardId << " of collection "
-                  << collectionId;
-            } else if (doc.operation == OperationType::kDropShard) {
-              auto const& collectionId = doc.collectionId;
-              self->_transactionHandler->abortTransactionsForShard(doc.shardId);
-              auto res = data.core->dropShard(doc.shardId, collectionId);
-              if (res.fail()) {
-                LOG_CTX("8c7a0", FATAL, self->loggerContext)
-                    << "Failed to drop shard " << doc.shardId
-                    << " of collection " << collectionId
-                    << " with error: " << res;
-                FATAL_ERROR_EXIT();
-              }
-              LOG_CTX("cdc44", TRACE, self->loggerContext)
-                  << "Dropped local shard " << doc.shardId << " of collection "
-                  << collectionId;
-            } else {
-              // Even though a shard was dropped before acquiring the snapshot,
-              // we could still see transactions referring to that shard.
-              if (!data.core->isShardAvailable(doc.shardId)) {
-                LOG_CTX("1970d", INFO, self->loggerContext)
-                    << "Will not apply transaction " << doc.tid << " for shard "
-                    << doc.shardId << " because it is not available";
-                continue;
-              }
-
-              auto res = self->_transactionHandler->applyEntry(doc);
-              if (res.fail()) {
-                LOG_CTX("1b08f", FATAL, self->loggerContext)
-                    << "Failed to apply entry " << entry->first
-                    << " to local shard " << doc.shardId
-                    << " with error: " << res;
-                FATAL_ERROR_EXIT();
-              }
-
-              if (doc.operation == OperationType::kAbortAllOngoingTrx) {
-                self->_activeTransactions.clear();
-                releaseIndex =
-                    self->_activeTransactions.getReleaseIndex(entry->first);
-              } else if (doc.operation == OperationType::kCommit ||
-                         doc.operation == OperationType::kAbort) {
-                self->_activeTransactions.erase(doc.tid);
-                releaseIndex =
-                    self->_activeTransactions.getReleaseIndex(entry->first);
-              } else {
-                self->_activeTransactions.emplace(doc.tid, entry->first);
-              }
+            auto [index, doc] = *entry;
```

Comment:
copy of document log entry might be expensive if we write huge batches.

Reply:
forget about that, the log entry contains a `ShardSlice` which is cheap to copy.